Add form philosophy (#385)
commercetools/ui-kit: philosophy/FORMS.md

# Forms

Most views in web applications are either lists or forms. So it's important to get forms right. It's important to offer good UX, but it's also important to make building and modifying forms straightforward for developers. Offering a consistent solution helps both.

With our approach to forms, we strive to

- **eliminate architectural discrepancy between different forms** by using the same approach for all forms

> - **offer good UX** by embedding the same principles for all forms in our solution

Reviewer: I think we can reword these two points to make it easier to read

> This document tries to outline the reasoning behind our form component design by showing the challenges we faced which led to the current design. This document aims to help contributors understand which considerations are relevant when building input components for ui-kit.

Reviewer: maybe form component ~~design~~ philosophy?

> It is not intended as a usage-guide of our input components. In fact it doesn't touch our input components for the most part.

Reviewer: maybe we should rename the page then?

How we design inputs for forms?

    - [The response shape of the server is inconvenient to build a form on](#the-response-shape-of-the-server-is-inconvenient-to-build-a-form-on)
    - [Providing defaults (Unifying inconsisent data: empty strings, `undefined` & `null`)](#providing-defaults-unifying-inconsisent-data-empty-strings-undefined--null)
    - [Handling numbers](#handling-numbers)
    - [Enriching received data (e.g. adding additional languages to localized strings)](#enriching-received-data-eg-adding-additional-languages-to-localized-strings)
  - [Validation](#validation)
    - [Translating error messages](#translating-error-messages)
    - [Multiple errors on a single field](#multiple-errors-on-a-single-field)
    - [Mapping async errors after submission back onto the form](#mapping-async-errors-after-submission-back-onto-the-form)
  - [Single inputs to determine multiple values (e.g. `MoneyInput`)](#single-inputs-to-determine-multiple-values-eg-moneyinput)
  - [Form values need to be shown in a user-friendly format](#form-values-need-to-be-shown-in-a-user-friendly-format)
  - [Preventing concurrent modifications](#preventing-concurrent-modifications)

## Typical steps of forms

> We want to offer a generic solution to forms. The following things usually need to be solved (roughly in that order) when building a form:

Reviewer: ~~things~~ problems?

Reviewer: in this order

> We want to offer a generic solution to forms. The following things usually need to be solved (roughly in that order) when building a form:

Reviewer 2: ```suggestion
We want to offer a generic solution to forms. The following problems usually need to be solved (roughly in this order) when building a form:
```

- fetching initial form data
- transforming the fetched data into form values
- modifying the form values draft based on the user input
- keeping track of modified fields to know which error messages to display
- validating the form as the user interacts with the form
- rendering validation errors
- validating the form on submission

> - Handling form submission

Reviewer: all the others are lowercase, this should be as well

```js
// conversions.js

// Converts document into form values
// This is used to transform the original data into form values.
export const docToFormValues = doc => ({
  id: doc.id,
  version: doc.version,
  name: doc.name,
});
```

Great, now we know that we're guaranteed a document which looks like this:

```
{
  id: String,
  version: Number,
  name: String,
}
```

### Modifying the form values draft based on the user input

We use [Formik](https://github.com/jaredpalmer/formik) to handle our form state. A very simple form could look like this:

```js
<GetProduct id="party-parrot">
  {product => (
    <Formik
      initialValues={docToFormValues(product)}
      render={formik => (
        <form onSubmit={formik.handleSubmit}>
          <input
            name="name"
            value={formik.values.name}
            onChange={formik.handleChange}
            onBlur={formik.handleBlur}
          />
          <button type="submit">Submit</button>
        </form>
      )}
    />
  )}
</GetProduct>
```

> As we've set the `name` attribute on the `input`, we can directly use `formik.handleChange` and `formik.handleBlur`. The `input` component will call these functions with an event, from `onChange` and `onBlur` respectively. That event will contain the input's name as `event.target.name`. Formik uses that information to change `formik.values.name` to the new value. It is therefore convenient when our input components call the change handlers with an event, and when that event containst the target.

Reviewer: 👍

Reviewer: typo on `containst`

Formik keeps track of the form values. It initializes the form values based on its `initialValues` property.

So at first Formik state looks something like this:

```js
{
  values: {
    id: 'party-parrot',
    version: 2,
    name: 'Party Parrot',
  },
  touched: {}
}
```

### Keeping track of modified fields to know which error messages to display

In the example above we also added `onBlur`. Just like with `onChange` Formik can read the information from the event passed to `onBlur` to know which form field to mark as touched.

As the user now makes changes to fields and blurs them, Formik's `touched` state gets updated:

```diff
{
  values: {
    id: 'party-parrot',
    version: 2,
    name: 'Party Parrot',
  },
  touched: {
+    name: true
  }
}
```

This indicates that the "name" field was touched.

> You'll notice that formik handles the `touched` state automatically based on `event.target`'s name. We never told it to set that value specifically. While a user is interacting with the a form, Formik produces a `touched` state solely based on the events `formik.handleBlur` receives.

Reviewer: Formik should be uppercased

    <Formik
      initialValues={docToFormValues(product)}
      validate={validate}
      render={formik => (
        <form onSubmit={formik.handleSubmit}>
          <input
            name="name"
            value={formik.values.name}
            onChange={formik.handleChange}
            onBlur={formik.handleBlur}
          />
          {formik.touched.name &&
            formik.errors.name &&
            formik.errors.name.missing && <div>A name is required</div>}
          <button type="submit">Submit</button>
        </form>
      )}
    />
  )}
</GetProduct>;
```

As `formik.touched.name` is only `true` after the user has left the name field, the errror message will not show up while the user is still interacting with the field (for the first time). This prevents scaring users by showing error messages too early.

The second time users change the input value of a field, the error message will go away as users interact with the field. This live validation helps users complete the form.

The ui-kit's `*Field` components have this behaviour baked into them, which you can see in _"Examples > Forms > Fields > Basic Form Example"_.

### Validating the form on submission

> In addition to the validation which runs while users fill out values, validation is also triggered when a user submits the form.

Reviewer: fill ~~out~~ in values

> In this case, there is another thing we need to be aware of: When a form is submitted, Formik automatically sets the `touched` state of all elements to `true`. This is done so that all error messages become visible. But Formik doesn't actually know about our input elements, it only knows about the stored values. So it tries to derive a `touched` state based on `formik.values`. This can lead to different `touched` states while a user interacts with a form and when the form is submitted. As we don't have any control over the `touched` state generated by Formik on submission, we need to ensure that the `touched` state we produce fits the one Formik will generate. Differences must be avoided to keep the code simple! Differing `touched` states lead to buggy code, unnecessary complexity and prop-type warnings.

Reviewer: 👍

- Too big of an API with `Form`, `withFormik`, `Field`, `ErrorMessage` and `Formik`. There are many ways to achieve the same thing. We generally try to use `<Formik />` only.
- The recommendation to generate error messages from the validation function mixes validation and rendering logic. Returning strings from the validation function gets in the way of using an applications' localization approach and makes it hard to determine failed validations in case an error needs to be shown in multiple places. It also makes it hard to attach multiple validation errors to a single field. That's why instead of returning a string for a field, we use objects holding booleans per field, e.g. `{ name: { missing: true } }`.

> ## Solutions recipies:

Reviewer: typo: recipes

When building a form, you might run into other challenges along the way. We already
have solutions for the most common cases which work with our approach to forms.
They are shown below:

> Here are some solution recipes for those cases:

Reviewer: these cases

The Formik state would then be

```js
{
  values: {
    id: 'party-parrot',
    version: 2,
    name: 'Party Parrot',
    inventory: 10,
  },
}
```

All browser inputs of type `number` will always only call `onChange` with either a valid number, or an empty string. `onChange` will never be called with an invalid number, `undefined` or a non-empty string.

This works for most cases, but we never get to see the original value entered by the user as the value gets parsed first. We either see a valid number or an empty string. This means we can't differntiate between `-0` and `0`. We also don't know wheter a user entered `9,8` (with a comma) or `9.8` (with a dot) as we will see the number `9.8` either way. We also can't differentiate between the user entering `00001` and `1.000`. We will see `1` in both cases. We also can't distinguish between `2e3` and `2000` as both are converted to `2000`. We also can't distinguish between the invalid `4e`, just `e` and an empty input - as we'll be called with `""` for all invalid values.

> The examples above were just for Google Chrome which prevents typing non-numbers. In Firefox it's possible to enter any string into `<input type="text" />` input elements. So it could happen that a user enters "hello" into a numeric input - and our Form thinks the user entered nothing, as Firefox will call `onChange` with an event holding an empty string.

Author: Nice catch!

    de: 'Party Papagei',
    es: 'Loro fiesta',
    ru: '',
  },
}
```

Notice that we're now passing additional information (`projectLocales`) to `docToFormValues`. This is a new pattern which allows us to enrich the recieved document. This approach of passing additional information to `docToFormValues` can be used to enrich form values in many ways.

### Validation

#### Translating error messages

During validation, we create error objects like this:

```js
{
  values: { name: '' },
  touched: { name: true },
  errors: { name: { missing: true } }
}
```

As you can see, we use boolean indicators to mark errors for form fields.
This allows us to determine how to render an error message from the render function,
instead of having to determine it within the validation function.

For example, we can do

```js
<Formik
  render={formik => (
    <div>
      <input
        type="text"
        name="name"
        onChange={handleChange}
        value={formik.values.name}
      />
      {formik.touched &&
        formik.touched.name &&
        formik.errors &&
        formik.errors.name &&
        formik.errors.name.missing && <div>Required field</div>}
    </div>
  )}
/>
```

> > Sidenote: Using the [optional chaining proposal (`?.`)](https://github.com/tc39/proposal-optional-chaining), we can shorten this code to

Reviewer: not sure if we really need to include this here, this doc is already long

Author: Yeah. It felt a bit weird to say this is a nice API when users have to do so many checks before accessing. That's why I wanted to point it out.

>
> ```js
> <Formik
>   render={formik => (
>     <div>
>       <input
>         type="text"
>         name="name"
>         onChange={handleChange}
>         value={formik.values.name}
>       />
>       {formik.touched?.name && formik.errors?.name?.missing && (
>         <div>Required field</div>
>       )}
>     </div>
>   )}
> />
> ```

We can now use any translation approach without having to intertwine it with validation:

```js
<Formik
  render={formik => (
    <div>
      <input
        type="text"
        name="name"
        onChange={handleChange}
        value={formik.values.name}
      />
      {formik.touched?.name && formik.errors?.name?.missing && (
        <FormattedMessage {...messages.requiredNameError} />
      )}
    </div>
  )}
/>
```

> Having this control also allows us to render arbitraryily complex error messages right from our code.

Reviewer: typo: arbitrarily

The value produced by `MoneyInput` is represented as `{ currencyCode: 'EUR', amount: '2.00' }`

> Sidenote: As the `MoneyInput` uses monetary formatting, we are using a text-input. This allows us to format values properly (e.g. by adding the trailing ".00") without browsers messing with the input. It also allows us to use decimal separators, e.g. for "1,000.00".

We can construct our Formik value like this:

```js
{
  id: 'party-parrot',
  version: 1,
  name: 'Party Parrot',
  price: { currencyCode: 'EUR', amount: '2.00' }
}
```

How do we now build a single input which receives `{ currencyCode: 'EUR', amount: '2.00' }` as its value and is able to change both using Formik's `handleChange`? We want to be able to render `<MoneyInput onChange={handleChange} onBlur={handleBlur} />` to make it as straightforward as using `<input type="text" />`.

There are two approaches:

- We can ensure `MoneyInput` always calls `onChange` with a full value like `{ currencyCode: 'EUR', amount: '2.00' }`
- We can adapt the input names to be `price.currencyCode` or `price.amount` respectively. Formik will read that name and set the nested value only.

> If we were to go with the first approach, we'd need to fake `event.target.name` in `onBlur` as well, otherwise formik would set `{ touched: { price: true } }` instead of `{ touched: { price: { amount: true, currencyCode: true } } }`. That is bad, as Formik will set ``{ touched: { price: { amount: true, currencyCode: true } } }`once the form is submitted - so we end up with a`touched`state which changes shape! There is a CodeSandbox which illustrates this problem. Try changing the price and keep an eye on the`touched`state. Then submit the form, and you'll see a different`touched` state.

Reviewer 2: ```suggestion
If we were to go with the first approach, we'd need to fake `event.target.name` in `onBlur` as well, otherwise formik would set `{ touched: { price: true } }` instead of `{ touched: { price: { amount: true, currencyCode: true } } }`. That is bad, as Formik will set `{ touched: { price: { amount: true, currencyCode: true } } }`once the form is submitted - so we end up with a `touched` state which changes shape! There is a CodeSandbox which illustrates this problem. Try changing the price and keep an eye on the `touched` state. Then submit the form, and you'll see a different `touched` state.
```

Reviewer 2: (the suggestion above fixes some bad formatting with backticks)

If we were to accept changes to the shape of our `touched` state, we'd need to tak the different shapes into consideration every time we work with the `touched` shape and it just generally makes things harder to follow.

So it's better to go with the second approach right away. We amend the names of the inputs by adding the `.amount` and `.currencyCode` suffix to their names and ids. When a parent doesn't pass in a name/id, we omit the property from the underlying input. This has solved that problem nicely for us.

### Form values need to be shown in a user-friendly format

As mentioned above, our `MoneyInput` consists of two parts: an currency selector and an amount input.

> We want the amount to be formatted like a monetary value, e.g. "200.00" or "2,000.00". However, the data we receive from our fantasy API looks like this: `{ currencyCode: 'EUR', centAmount: 200 }` for 2€.

Reviewer 2: ```suggestion
We want the amount to be formatted like a monetary value, e.g. "200.00" or "2,000.00". However, the data we receive from our API looks like this: `{ currencyCode: 'EUR', centAmount: 200 }`, which means **2€**.
```

Reviewer 2: "Fantasy" just sounds weird in this context. A better word might be "fictitious" if you want to point out that it's not a real API.

Author: Yeah that's better! Done 👍

Reviewer: Yeah I think the ~best~ english term would be hypothetical API 😆 :trollface:

How do we go about formatting the value properly? As our form values always hold the actual value entered by the user, we need to format the initial value in `docToFormValues` already.

Our ideal format should:

- use separators based on the users locale (`2.000,00` vs `2,000.00`)
- apply a number of fraction digits based on the currency as some currencies use more/less than 2 fraction digits (`2,000.00` vs `2,000.000`)

A `locale` defines which separators to use. The money value itself already contains a `currencyCode`, which we can use to determine the number of fraction digits.

> To format the value, we thus only need to pass a `locale` to `docToFormValues`:

Reviewer 2: ```suggestion
Therefore, to format the value we only need to pass a `locale` to `docToFormValues`:
```
